elf_reader.go

```diff
@@ -217,38 +250,128 @@ func (ec *elfCode) loadMaps(mapSections map[elf.SectionIndex]*elf.Section) (map[
 		for i, offset := 0, uint64(0); i < len(syms); i, offset = i+1, offset+size {
 			mapSym := syms[offset]
 			if mapSym == "" {
-				fmt.Println(syms)
-				return nil, errors.Errorf("section %s: missing symbol for map at offset %d", sec.Name, offset)
+				return errors.Errorf("section %s: missing symbol for map at offset %d", sec.Name, offset)
 			}
 
 			if maps[mapSym] != nil {
-				return nil, errors.Errorf("section %v: map %v already exists", sec.Name, mapSym)
+				return errors.Errorf("section %v: map %v already exists", sec.Name, mapSym)
 			}
 
 			lr := io.LimitReader(r, int64(size))
 
 			var spec MapSpec
 			switch {
 			case binary.Read(lr, ec.ByteOrder, &spec.Type) != nil:
-				return nil, errors.Errorf("map %v: missing type", mapSym)
+				return errors.Errorf("map %v: missing type", mapSym)
 			case binary.Read(lr, ec.ByteOrder, &spec.KeySize) != nil:
-				return nil, errors.Errorf("map %v: missing key size", mapSym)
+				return errors.Errorf("map %v: missing key size", mapSym)
 			case binary.Read(lr, ec.ByteOrder, &spec.ValueSize) != nil:
-				return nil, errors.Errorf("map %v: missing value size", mapSym)
+				return errors.Errorf("map %v: missing value size", mapSym)
 			case binary.Read(lr, ec.ByteOrder, &spec.MaxEntries) != nil:
-				return nil, errors.Errorf("map %v: missing max entries", mapSym)
+				return errors.Errorf("map %v: missing max entries", mapSym)
 			case binary.Read(lr, ec.ByteOrder, &spec.Flags) != nil:
-				return nil, errors.Errorf("map %v: missing flags", mapSym)
+				return errors.Errorf("map %v: missing flags", mapSym)
 			}
 
 			if _, err := io.Copy(internal.DiscardZeroes{}, lr); err != nil {
-				return nil, errors.Errorf("map %v: unknown and non-zero fields in definition", mapSym)
+				return errors.Errorf("map %v: unknown and non-zero fields in definition", mapSym)
 			}
 
 			maps[mapSym] = &spec
 		}
 	}
-	return maps, nil
+
+	return nil
+}
+
+func (ec *elfCode) loadBTFMaps(maps map[string]*MapSpec, mapSections map[elf.SectionIndex]*elf.Section, spec *btf.Spec) error {
+
+	if spec == nil {
+		return errors.Errorf("missing BTF")
+	}
+
+	for idx, sec := range mapSections {
+		syms := ec.symbolsPerSection[idx]
+		if len(syms) == 0 {
+			return errors.Errorf("section %v: no symbols", sec.Name)
+		}
+
+		for _, sym := range syms {
+			if maps[sym] != nil {
+				return errors.Errorf("section %v: map %v already exists", sec.Name, sym)
+			}
+
+			btfMap, err := spec.Map(sym)
+			if err != nil {
+				return errors.Wrap(err, "can't get BTF")
+			}
+
+			var mapType, flags, maxEntries uint32
+			for _, member := range btfMap.Type().Members {
+				switch member.Name {
+				case "type":
+					mapType, err = uintFromBTF(member.Type)
+					if err != nil {
+						return errors.Wrapf(err, "map %s: can't get type", sym)
+					}
+
+				case "map_flags":
+					flags, err = uintFromBTF(member.Type)
+					if err != nil {
+						return errors.Wrap(err, "can't get BTF map flags")
```

Comment:
For the type, it provides the map symbol yet for the flags here there's no map symbol, should these all consistently add the map symbol so that users would have some idea which part of the BTF is messed up?

Reply:
Fixed.